allennlp/training/event_based_trainer/event_based_trainer.py

```diff
@@ -0,0 +1,87 @@
+"""
+blah
+"""
+from typing import Dict, Any, Union, List, Iterable, Callable
+from collections import defaultdict
+
+from allennlp.common.tqdm import Tqdm
+from allennlp.data.iterators.data_iterator import TensorDict
+from allennlp.models.model import Model
+from allennlp.training.event_based_trainer.callback import Callback
+from allennlp.training.event_based_trainer.events import Events
+from allennlp.training.trainer_base import TrainerBase
+
+
+class EventBasedTrainer(TrainerBase):
+    def __init__(self,
+                 model: Model,
+                 epochs_generator: Iterable[Iterable[TensorDict]],
+                 serialization_dir: str,
+                 cuda_device: Union[int, List] = -1) -> None:
+        super().__init__(serialization_dir, cuda_device)
+        self.model = self._move_to_gpu(model)
+        self.epochs_generator = epochs_generator
+        self.callbacks: Dict[str, List[Callback]] = defaultdict(list)
```

Comment:
You should be able to pass a dictionary of callbacks to the constructor, so you can specify all of this `from_params`.

Reply:
:+1: